logical/plugin/backend_server.go

```diff
@@ -0,0 +1,146 @@
+package plugin
+
+import (
+	"net/rpc"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vault/logical"
+)
+
+type backendPluginServer struct {
+	broker  *plugin.MuxBroker
+	backend logical.Backend
+}
+
+func (b *backendPluginServer) HandleRequest(args *HandleRequestArgs, reply *HandleRequestReply) error {
+	conn, err := b.broker.Dial(args.StorageID)
+	if err != nil {
+		*reply = HandleRequestReply{
+			Error: plugin.NewBasicError(err),
+		}
+		return nil
+	}
+	c := rpc.NewClient(conn)
+	defer c.Close()
+
+	storage := &StorageClient{client: c}
+	args.Request.Storage = storage
+
+	resp, err := b.backend.HandleRequest(args.Request)
+	*reply = HandleRequestReply{
+		Response: resp,
+		Error:    plugin.NewBasicError(err),
+	}
+
+	return nil
+}
+
+func (b *backendPluginServer) SpecialPaths(_ interface{}, reply *SpecialPathsReply) error {
+	*reply = SpecialPathsReply{
+		Paths: b.backend.SpecialPaths(),
+	}
+	return nil
+}
+
+func (b *backendPluginServer) Logger(_ interface{}, reply *BackendLoggerReply) error {
+	*reply = BackendLoggerReply{
+		Logger: b.backend.Logger(),
+	}
+	return nil
+}
+
+func (b *backendPluginServer) HandleExistenceCheck(args *HandleExistenceCheckArgs, reply *HandleExistenceCheckReply) error {
+	conn, err := b.broker.Dial(args.StorageID)
+	if err != nil {
+		*reply = HandleExistenceCheckReply{
+			Error: plugin.NewBasicError(err),
+		}
+		return nil
+	}
+	c := rpc.NewClient(conn)
+	defer c.Close()
+
+	storage := &StorageClient{client: c}
+	args.Request.Storage = storage
+
+	checkFound, exists, err := b.backend.HandleExistenceCheck(args.Request)
+	*reply = HandleExistenceCheckReply{
+		CheckFound: checkFound,
+		Exists:     exists,
+		Error:      plugin.NewBasicError(err),
+	}
+
+	return nil
+}
+
+func (b *backendPluginServer) Cleanup(_ interface{}, _ *struct{}) error {
+	b.backend.Cleanup()
+	return nil
+}
+
+func (b *backendPluginServer) Initialize(_ interface{}, _ *struct{}) error {
+	err := b.backend.Initialize()
+	return err
+}
+
+func (b *backendPluginServer) InvalidateKey(args string, _ *struct{}) error {
+	b.backend.InvalidateKey(args)
+	return nil
+}
+
+func (b *backendPluginServer) Configure(args *ConfigureArgs, reply *ConfigureReply) error {
```

Comment:
I wonder if it makes more sense to name this function Factory and have it actually instantiate the backend object by calling the passed in backend.Factory method? That way there is no divergence in how builtin backends and plugin backends configure themselves. In both cases it could happen in the Factory method.

That would mean the backendPluginServer holds the Factory() function and the backend object. 

What do you think?